router/http.go

```diff
@@ -356,310 +345,27 @@ type httpService struct {
-		if a == addr {
-			ok = true
-			break
-		}
-	}
-	if !ok {
-		return ""
-	}
-
-	return addr
-}
-
-func (s *httpService) newStickyCookie(backend string) *http.Cookie {
-	var nonce [24]byte
-	_, err := io.ReadFull(rand.Reader, nonce[:])
-	if err != nil {
-		panic(err)
-	}
-	out := make([]byte, len(nonce), len(nonce)+len(backend)+secretbox.Overhead)
-	copy(out, nonce[:])
-	out = secretbox.Seal(out, []byte(backend), &nonce, s.cookieKey)
-
-	return &http.Cookie{Name: stickyCookie, Value: base64.StdEncoding.EncodeToString(out), Path: "/"}
+	proxy proxy.Proxy
 }
 
 func (s *httpService) ServeHTTP(w http.ResponseWriter, req *http.Request) {
 	req.Header.Set("X-Request-Start", strconv.FormatInt(time.Now().UnixNano()/int64(time.Millisecond), 10))
 	req.Header.Set("X-Request-Id", random.UUID())
 
-	addrs := shuffle(s.sc.Addrs())
-	if len(addrs) == 0 {
-		log.Println("no backends found")
-		fail(w, 503)
-		return
-	}
-
-	isSticky := false
-	stickyAddr := ""
-	if req.Header.Get(hdrUseStickySessions) == "true" {
-		// TODO(bgentry): switch to better way to check sticky setting
-		isSticky = true
-		if stickyAddr = s.stickyCookieAddr(req); stickyAddr != "" {
-			sortStringFirst(addrs, stickyAddr)
-		}
-	}
-	req.Header.Del(hdrUseStickySessions) // delete this no matter what
-
-	// Most of this is borrowed from httputil.ReverseProxy
-	outreq := &http.Request{}
-	*outreq = *req // includes shallow copies of maps, but okay
-
 	// Pass the Request-URI verbatim without any modifications
-	outreq.URL.Opaque = strings.Split(strings.TrimPrefix(req.RequestURI, req.URL.Scheme+":"), "?")[0]
-	outreq.URL.Scheme = "http"
-	outreq.Proto = "HTTP/1.1"
-	outreq.ProtoMajor = 1
-	outreq.ProtoMinor = 1
-	outreq.Close = false
-
-	// TODO: Proxy HTTP CONNECT? (example: Go RPC over HTTP)
-
-	// Directly bridge `Connection: Upgrade` requests
-	if strings.ToLower(outreq.Header.Get("Connection")) == "upgrade" {
-		s.forwardAndProxyTCP(w, outreq, addrs, stickyAddr, isSticky)
-		return
-	}
+	req.URL.Opaque = strings.Split(strings.TrimPrefix(req.RequestURI, req.URL.Scheme+":"), "?")[0]
 
-	// Remove hop-by-hop headers to the backend.  Especially
-	// important is "Connection" because we want a persistent
-	// connection, regardless of what the client sent to us.  This
-	// is modifying the same underlying map from req (shallow
-	// copied above) so we only copy it if necessary.
-	copiedHeaders := false
-	for _, h := range hopHeaders {
-		if outreq.Header.Get(h) != "" {
-			if !copiedHeaders {
-				outreq.Header = make(http.Header)
-				copyHeader(outreq.Header, req.Header)
-				copiedHeaders = true
-			}
-			outreq.Header.Del(h)
-		}
-	}
-
-	var (
-		backend string
-		res     *http.Response
-		err     error
-	)
-
-	for _, backend = range addrs {
-		// TODO: limit number of backends tried
-		// TODO: temporarily quarantine failing backends
+	//TODO(benburkert): only call this when the ss set is updated, instead of every request
+	s.proxy.UpdateBackends(s.sc.Addrs())
```

Comment:
I'm guessing it's going to be quite expensive to pass around the full list of backends every time one changes. But even more importantly, we already have that information in memory on another struct. It'd be better if we didn't have to duplicate the lists of backends on the proxy, and instead could look them up directly on the thing that's already storing them.

Ultimately that'd give us fewer copies of data to have to keep in sync, and fewer locks to deal with.